sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala

```diff
@@ -39,15 +40,19 @@ case class BatchScanExec(
     keyGroupedPartitioning: Option[Seq[Expression]] = None,
     ordering: Option[Seq[SortOrder]] = None,
     @transient table: Table,
-    commonPartitionValues: Option[Seq[InternalRow]] = None) extends DataSourceV2ScanExecBase {
+    commonPartitionValues: Option[Seq[(InternalRow, Int)]] = None,
+    applyPartialClustering: Boolean = false,
+    replicatePartitions: Boolean = false) extends DataSourceV2ScanExecBase {
```

Comment:
This happens quite a few times that we add a new parameter to `BatchScanExec` and change a lot of code which does a pattern matching over `BatchScanExec`. Can we put SPJ-related information in a wrapper class? Then `BatchScanExec` can be more stable. cc @sunchao

Reply:
@cloud-fan yes, it's a good idea. The issue is quite annoying. Let me create a PR for it.